[SPARK-45713][PYTHON] Support registering Python data sources

allisonwang-db opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
This PR adds support for registering Python data sources.
Users can register a Python data source using the class:

```python
class MyDataSource(DataSource):
    ...

spark.dataSource.register(MyDataSource)
```

This will allow users to use the data source using its name (to be supported in SPARK-45639)
```python
spark.read.format("MyDataSource").load()
```

The data sources registered are stored in `sharedState` and can be accessed by all spark sessions.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
To support Python data source.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
Yes. This PR adds a new API in spark session to support registering data sources.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->
Unit tests.

### Was this patch authored or co-authored using generative AI tooling?
<!--
If generative AI tooling has been used in the process of authoring this patch, please include the
phrase: 'Generated-by: ' followed by the name of the tool and its version.
If no, write 'No'.
Please refer to the [ASF Generative Tooling Guidance](https://www.apache.org/legal/generative-tooling.html) for details.
-->
No

allisonwang-db committed:
data source registry

allisonwang-db commented:
cc @HyukjinKwon

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceRegistry.scala:
```diff
@@ -0,0 +1,77 @@
+import org.apache.spark.sql.util.CaseInsensitiveStringMap
+
+
+trait DataSourceRegistry {
```
I think we should somehow reuse `DataSourceRegister`.

HyukjinKwon replied:
cc @cloud-fan

cloud-fan replied:
These two are completely different. We should probably rename this one here.

allisonwang-db replied:
Yea we should pick a different name. The name "registry" is taken from "FunctionRegistry" and "TableFunctionRegistry". How about "DataSourceStore" or "DataSourceManager"

allisonwang-db committed:
fix tests

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/DataSourceRegistration.scala:
```diff
@@ -0,0 +1,50 @@
+ * Functions for registering user-defined data sources.
+ * Use `SparkSession.dataSource` to access this.
+ *
+ * @since 4.0.0
```
do we need to expose it now? It's only called by PySpark internally. We can mark it as `private[sql]`

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala:
```diff
@@ -228,6 +228,11 @@ class SparkSession private(
+  /**
+   * A collection of methods for registering user-defined data sources.
+   */
+  def dataSource: DataSourceRegistration = sharedState.dataSourceRegistration
```
ditto

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceRegistry.scala:
```diff
@@ -0,0 +1,77 @@
+  override def clone(): DataSourceRegistry = throw new CloneNotSupportedException()
+}
+
+class SimpleDataSourceRegistry extends DataSourceRegistry with Logging {
```
Let's not add unnecessary abstraction. Why would we need different implementations of `DataSourceRegistry`? Isn't it just a mapping between name and builder?

allisonwang-db replied:
Sounds good. Changed this to a simple class with a concurrent hasmap.

allisonwang-db committed:
address comments

allisonwang-db commented:
@HyukjinKwon @cloud-fan Thanks for the review! I've addressed the comments. Note this PR does not include the data source lookup logic. I will work on it in the next PR ([SPARK-45639](https://issues.apache.org/jira/browse/SPARK-45639)) (and also address any additional comments)

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/DataSourceRegistration.scala:
```diff
@@ -0,0 +1,48 @@
+ * Functions for registering user-defined data sources.
+ * Use `SparkSession.dataSource` to access this.
+ */
+@Evolving
```
nit: we don't need it for private APIs.

cloud-fan replied:
We can fix it in the followup

cloud-fan commented:
thanks, merging to master!

HyukjinKwon commented:
Hey, I don't think we should go ahead. It should have a lookup logic first, and then we should think about something extra runtime registration

panbingkun commented:
After this pr, failed to run 'SparkThrowableSuite' in GA, eg:
https://github.com/panbingkun/spark/actions/runs/6702704679/job/18212102847
<img width="857" alt="image" src="https://github.com/apache/spark/assets/15246973/9538d6c8-1d75-452c-b72c-567f90ae9f14">
I have submitted a PR to fix this issue: https://github.com/apache/spark/pull/43598
